### PR TITLE
No locking ReferenceCountedBuffer

### PR DESCRIPTION
**Pull request by Scooletz**
`ReferenceCountedBuffer` now uses `Interlocked` operations instead of locks.

### CODE REVIEW

File: src/Channels/ReferenceCountedBuffer.cs [hunk 1]
```diff
@@ -28,19 +25,16 @@ public IBuffer Preserve(int offset, int count, out int newStart, out int newEnd)
 
         public void Dispose()
         {
-            lock (_lockObj)
```
Comment (davidfowl) [inline src/Channels/ReferenceCountedBuffer.cs:31]: This was originally an interlocked but I believe the decrement and dispose need to be atomic.

Comment (Scooletz) [reply src/Channels/ReferenceCountedBuffer.cs:31]: I can see it in the history now (I didn't take a look before), but I see no reason for lock :)

Comment (davidfowl) [reply src/Channels/ReferenceCountedBuffer.cs:31]: In theory, this buffer can be leasable with the wrong reference count after DisposeBuffer is called and before the count is reset to 1.

Comment (Scooletz) [reply src/Channels/ReferenceCountedBuffer.cs:31]: I see it now. I'm retracting this PR.

**Pull request closed without merging**